### PR TITLE
Annotations redesign

### PR DESCRIPTION
**Pull request by Pazus**
Redesign according to #133 

Changin code to use new `%displayname` annotation a realized I like the old way `%suite(Name of the suite)`, `%test(Name of the test)` really much, so maybe we should get back to the old approach.
Additionaly as IDE doesn't help us with autocomplete on annotations, the less of them we need to specify, the more stable is test package configuration.
What do you think? Look at the comparison of old and new examples.

Or we can use both aproaches with `%displayname` prioritized.

### CODE REVIEW

File: docs/userguide/annotations.md [hunk 1]
```diff
@@ -1,65 +1,68 @@
...
-%testteardown | Marks that special teardown procedure has to be run before the test instead of the default one
-%suitesetup | Procedure with executes once at the beginning of the suite and doesn't executes before each test
-%suiteteardown | Procedure with executes once after the execution of the last test in the suite.
+| Annotation |Level| Describtion |
+| --- | --- | --- |
+| `%suite` | Package | Marks package to be a suite of tests This way all testing packages might be found in a schema. |
+| `%suitepath(<path>)` | Package | Similar to java package. The annotation allows logical grouping of suites into hierarcies. |
+| `%displayname(<description>)` | Package/procedure | Human-familiar describtion of the suite/test. `%displayname(Name of the suite/test)` |
+| `%test` | Procedure | Denotes that a method is a test method. |
+| `%beforeall` | Procedure | Denotes that the annotated procedure should be executed before all `%test` methods in the current suite. |
```
Comment (jgebal) [inline docs/userguide/annotations.md]: Denotes that the annotated procedure should be executed **before all elements of the current suite**.
If `%beforeall` is defined more then once, the last `%beforeall` in the suite is used.

File: docs/userguide/annotations.md [hunk 2]
```diff
@@ -1,65 +1,68 @@
...
+| `%afterall` | Procedure | Denotes that the annotated procedure should be executed after all `%test` methods in the current suite. |
+| `%beforeeach` | Procedure | Denotes that the annotated procedure should be executed before each `%test` method in the current suite. |
```
Comment (jgebal) [inline docs/userguide/annotations.md]: Each element of suite. We consider sub-suites on a suitepath. If we have suitepath with parent suite package and child suite package, the `%beforeeach` of parent suite will fire before invoking the child suite.

**Comment by jgebal** [comment]
I also like the idea of keeping the annotations as simple as possible.
I'd also like to come back to the `%context` or `%subsuite` idea, where we can nest structures like this:
```
create or replace package demo_equal_matcher as

  -- %suite(Equal matcher)
  -- %suitepackage(org.utplsql.v3.demo.matchers)

  -- %context
  -- %displayname(Comparing user defined types for equality)

    -- %test(Gives success when comparing identical objects containing identical data)
    procedure object_compare_success;

    -- %test(Gives failure when comparing to a null actual)
    procedure object_compare_null_actual;

    -- %test(Gives failure when comparing to a null expected)
    procedure object_compare_null_expected;

    -- %test(Gives success when comparing null actual to a null expected)
    procedure object_compare_null_both_ok;

    -- %test(Gives failure when comparing null actual to a null expected, setting null equal to false)
    procedure object_compare_null_both_fail;

    -- %test(Gives failure when comparing identical objects containing different data)
    procedure object_compare_different_data;

    -- %test(Gives failure when comparing different objects containing identical data)
    procedure object_compare_different_type;

  -- %end_context

end;
/
```

I think that `%suitepath` annotation combined with `%context`/`%endcontext` give huge flexibility and incredible power to the framework.

**Comment by jgebal** [comment]
As we are on the renaming of annotations.
Should we keep the annotations like:
`%suitepath` or rather name it `%suite_path`
Also, should annotation names be case sensitive or not? I think not, as PLSQL is not case-sensitive.
Regardless of the decision on case-sensitivity, we need to explicitly state this in the documentation.

File: docs/userguide/annotations.md (hunk 2 above)
Comment (Pazus) [inline docs/userguide/annotations.md]: I'm not sure it is what we want... I'd stay with beforeeach/aftereach executes only for tests.

File: docs/userguide/annotations.md (hunk 1 above)
Comment (Pazus) [inline docs/userguide/annotations.md]: agree

**Comment by Pazus** [comment]
1. I've implemented not all annotations from the list of #133 , Nested and tags will be ready in 3.1
2. I like `suitepath` more, it's less error prone in typing as we dont have compilation spell checker. @rlove, @Shoelace what do you think?
3. annotations are case-insensitive

**Comment by jgebal** [comment]
Can you add all that info to the readme so we have no doubts in the future?
- Annotations are case insensitive.
- If an annotation is defined more then once, the last one in scope of item (suite/test) is used.

**Comment by Shoelace** [comment]
in regards to  %suitepath or %suite_path   i would use not use the underscore and have it as %suitepath

**Comment by jgebal** [comment]
@pazus, can you put strings 'suite' etc. into constants?

**Pull request merged**